x/mobileproxy/mobileproxy.go

```diff
@@ -90,3 +90,94 @@ func RunProxy(localAddress string, transportConfig string) (*Proxy, error) {
 	}
 	return &Proxy{host: host, port: port, server: server}, nil
 }
+
+// StreamDialer encapsulates the logic to create stream connections (like TCP).
+type StreamDialer struct {
+	transport.StreamDialer
+}
+
+// NewStreamDialerFromConfig creates a [StreamDialer] based on the given config.
+// The config format is specified in https://pkg.go.dev/github.com/Jigsaw-Code/outline-sdk/x/config#hdr-Config_Format.
+func NewStreamDialerFromConfig(transportConfig string) (*StreamDialer, error) {
+	dialer, err := config.NewStreamDialer(transportConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &StreamDialer{dialer}, nil
+}
+
+// LogWriter is used as a sink for logging.
+type LogWriter io.StringWriter
+
+// Adaptor to convert an [io.StringWriter] to a [io.Writer].
+type stringToBytesWriter struct {
+	w io.Writer
+}
+
+// WriteString implements [io.StringWriter].
+func (w *stringToBytesWriter) WriteString(logText string) (int, error) {
+	return io.WriteString(w.w, logText)
+}
+
+// NewStderrLogWriter creates a [LogWriter] that writes to the standard error output.
+func NewStderrLogWriter() LogWriter {
+	return &stringToBytesWriter{os.Stderr}
+}
+
+// Adaptor to convert an [io.Writer] to a [io.StringWriter].
+type bytestoStringWriter struct {
+	sw io.StringWriter
+}
+
+// Write implements [io.Writer].
+func (w *bytestoStringWriter) Write(b []byte) (int, error) {
+	return w.sw.WriteString(string(b))
+}
+
+func toWriter(logWriter LogWriter) io.Writer {
+	if logWriter == nil {
+		return nil
+	}
+	if w, ok := logWriter.(io.Writer); ok {
+		return w
+	}
+	return &bytestoStringWriter{logWriter}
+}
+
+// NewSmartStreamDialer automatically selects a DNS and TLS strategy to use, and return a [StreamDialer]
+// that will use the selected strategy.
+// It uses testDomain to find a strategy that works when accessing those domains.
+// The strategies to search are given in the searchConfig. An example can be found in
+// https://github.com/Jigsaw-Code/outline-sdk/x/examples/smart-proxy/config.json
+func NewSmartStreamDialer(testDomains *StringList, searchConfig string, logWriter LogWriter) (*StreamDialer, error) {
+	logBytesWriter := toWriter(logWriter)
+	// TODO: inject the base dialer for tests.
+	finder := smart.StrategyFinder{
+		LogWriter:    logBytesWriter,
+		TestTimeout:  5 * time.Second,
+		StreamDialer: &transport.TCPDialer{},
+		PacketDialer: &transport.UDPDialer{},
+	}
+	testDomainsSlice := append(make([]string, 0, len(testDomains.list)), testDomains.list...)
+	dialer, err := finder.NewDialer(context.Background(), testDomainsSlice, []byte(searchConfig))
+	if err != nil {
+		return nil, fmt.Errorf("failed to find dialer: %v", err)
+	}
+	return &StreamDialer{dialer}, nil
+}
+
+// StringList allows us to pass a list of strings to the Go Mobile functions, since Go Mobiule doesn't
```

Comment:
Go Mobile seems to be a very different thing if you search it on google.

```suggestion
// StringList allows us to pass a list of strings to the Go Mobile functions, since Gomobile doesn't
```